opnode/rollup/driver/step.go

```diff
@@ -0,0 +1,298 @@
+package driver
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"math/big"
+	"time"
+
+	"github.com/ethereum-optimism/optimistic-specs/opnode/eth"
+	"github.com/ethereum-optimism/optimistic-specs/opnode/l2"
+	"github.com/ethereum-optimism/optimistic-specs/opnode/rollup"
+	"github.com/ethereum-optimism/optimistic-specs/opnode/rollup/derive"
+
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/common/hexutil"
+	"github.com/ethereum/go-ethereum/core/types"
+	"github.com/ethereum/go-ethereum/log"
+)
+
+type outputImpl struct {
+	dl     Downloader
+	l2     Engine
+	log    log.Logger
+	Config rollup.Config
+}
+
+func (d *outputImpl) createNewBlock(ctx context.Context, l2Head eth.L2BlockRef, l2SafeHead eth.BlockID, l2Finalized eth.BlockID, l1Origin eth.L1BlockRef) (eth.L2BlockRef, *derive.BatchData, error) {
+	d.log.Info("creating new block", "l2Head", l2Head)
+
+	// If the L1 origin changed this block, then we are in the first block of the epoch
+	firstEpochBlock := l2Head.L1Origin.Number != l1Origin.Number
+
+	fetchCtx, cancel := context.WithTimeout(ctx, time.Second*20)
+	defer cancel()
+
+	var l1Info derive.L1Info
+	var receipts types.Receipts
+	var err error
+	// Include deposits if this is the first block of an epoch
+	if firstEpochBlock {
+		l1Info, _, receipts, err = d.dl.Fetch(fetchCtx, l1Origin.Hash)
+	} else {
+		l1Info, err = d.dl.InfoByHash(fetchCtx, l1Origin.Hash)
+		// don't fetch receipts if we do not process deposits
+	}
+	if err != nil {
+		return l2Head, nil, fmt.Errorf("failed to fetch L1 block info of %s: %v", l1Origin, err)
+	}
+
+	l1InfoTx, err := derive.L1InfoDepositBytes(l2Head.Number+1, l1Info)
+	if err != nil {
+		return l2Head, nil, err
+	}
+	var txns []l2.Data
+	txns = append(txns, l1InfoTx)
+	deposits, err := derive.DeriveDeposits(l2Head.Number+1, receipts)
+	d.log.Info("Derived deposits", "deposits", deposits, "l2Parent", l2Head, "l1Origin", l1Origin)
+	if err != nil {
+		return l2Head, nil, fmt.Errorf("failed to derive deposits: %v", err)
+	}
+	txns = append(txns, deposits...)
+
+	depositStart := len(txns)
```

Comment:
The `depositStart` should be set before we append deposits on L62. Otherwise, the batch to be submitted will be empty (L87).

Reply:
Unless the point is to create batches of any additional transactions that the engine may have added. Not sure on the intent here.